include/circt/Scheduling/Scheduler.h

```diff
@@ -0,0 +1,253 @@
+//===- Scheduler.h - Common interface for scheduling algorithms -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Static scheduling algorithms for use in HLS flows all solve similar problems.
+// The classes in this file serve as an interface between clients and algorithm
+// implementations, and model a basic scheduling problem and some commonly used
+// extensions (e.g. modulo scheduling). This includes problem-specific utility
+// and verification methods.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SCHEDULING_SCHEDULER_H
+#define CIRCT_SCHEDULING_SCHEDULER_H
+
+#include "circt/Support/LLVM.h"
+
+#include "mlir/Support/StorageUniquer.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace circt {
+namespace sched {
+
+class Dependence;
+class OperatorType;
+
+/// This class models the most basic scheduling problem.
+///
+/// A problem instance is comprised of:
+///
+///  - *Operations*: The vertices in the data-flow graph to be scheduled. We use
+///    the abbreviation `op` in the scheduling API.
+///  - *Dependences*: The edges in the data-flow graph to be scheduled, modeling
+///    a precedence relation between the involved operations. Individual operand
+///    and result indices can be distinguished. Abbreviation: `dep`
+///  - *Operator types*: An abstraction of the characteristics of the target
+///    representation, e.g. modules in the  HLS component library, available
+///    functional units, etc. -- operations are executed on instances/units of
+///    an operator type. Abbreviation: `opr`
+///
+/// All components of the problem (ops, deps, oprs, as well as the problem
+/// itself) can be annotated with *properties*. In this basic problem, we model
+///
+/// - `assocOpr`, an operation-property that stores the operator type associated
+///   with each operation.
+/// - `latency`, an operator type-property denoting the number of time steps
+///   after which the operator's result is available.
+/// - `startTime`, an operation-property for the time step in which an operation
```

Comment:
Would it make sense for the output to be a [`PipelineOp`](https://github.com/llvm/circt/blob/012d69a27e4e11011c6c26ca9927a3a312af4b89/include/circt/Dialect/StaticLogic/StaticLogic.td#L32)?

Reply:
I think the interface should be dialect-agnostic at this level. Constructing a `PipelineOp` from the computed start times sounds like a great candidate for a utility function, if that becomes a common need for HLS flows in the future.

Reply:
> I think the interface should be dialect-agnostic at this level.

I meant to comment on this, I agree with @jopperm here. It _could_ make sense to use a PipelineOp, but this Scheduling.h infrastructure probably doesn't want to force that.

Reply:
Agreed